"cupid-clear" Command Implementation Addressing Issue #85

shivaniikum opened the pull request:
**This pull request addresses issue #85 and contains the following additions:** 
1) A new clear.py script located at "cupid/clear.py" 
2) cupid-clear = "cupid.clear:clear" added to the last line of the pyproject.toml file


**The clear.py script can be run with the following steps:**

1) After adding these changes to your local environment, type "pip install -e ." in your terminal

2) Navigate to the directory where the config.yml file is located (this config.file contains the run_dir variable which is the path to the computed_notebooks folder you want to clear). For example: "cd examples/coupled_model"

3) Run the command "cupid-clear config.yml" and this should clear the computed_notebooks folder at the path specified by the run_dir variable in your config.yml file (if you're running cupid-clear from the examples/coupled_model directory and the config.yml located here has run_dir= . then the folder at the path examples/coupled_model/computed_notebooks will be cleared, if it exists)

TeaganKing commented on cupid/clear.py:
```diff
@@ -0,0 +1,52 @@
+    except Exception as e:
+        print(f"Error occurred while clearing contents of the file at path {folderPath}: {e}")
+        
+def readConfigFile(configFilePath):
```
Within this function, it may be useful to keep a similar structure for reading the config file to that in `run.py`. So, instead of the `yaml.safe_load(file)`, you may want to use the pre-exisiting get_control_dict function as is used [here](https://github.com/shivaniikum/CUPiD/blob/7b2c7c4e31b1b2ff36d80e0510b739c60d17b1cb/cupid/run.py#L34) to read the contents of the configuration file, and then read in `control['data_sources']['run_dir']`.

TeaganKing reviewed (changes requested):
Thank you for contributing this feature! It looks like this is deleting all files under and including `computed_notebooks` when `cupid-clear config.yaml` is run, so it's working as expected, which is great!

When I set up the environment again from scratch, I was running into a few issues with the cupid-dev environment (not installing all packages properly with `mamba env create -f environments/dev-environment.yml`) that I'd like to double check once our machines are back up (hopefully soon).

One note, it may be useful to document this capability (just one line or so) in the README.md under the 'Running' section (before 'Cupid Options').

rmshkv reviewed (commented):
Agreed with Teagan's suggestions about the get_control_dict function and adding to the readme! I also want to test run this myself as soon as our machines are back up.

TeaganKing commented:
Also, just to be clear, I don't think any of my aforementioned environment issues are related to this PR since there were no changes to the environment yaml files-- I just wanted to be sure to note that and check to make sure we don't have other issues going on :)

shivaniikum commented:
Thank you for your feedback. I've modified clear.py to use the pre-exisiting get_control_dict function, then to read in control['data_sources']['run_dir']. I also added instructions about clear.py to the README.md. Please let me know if this works correctly.

TeaganKing commented:
Great! Thanks @shivaniikum . I'll test that again once our machines are back up!

TeaganKing reviewed (approved):
Ok, this is looking great! Thanks for making those changes.

The environment issues were specific to my machine and how I had set things up-- I've also tested on derecho (an NCAR machine) and everything looks good with the environment setup now.

The new method of finding the configuration file `run_dir` seems to be working smoothly.

I'll approve this PR, but I think @rmshkv wanted to test it out as well before we merge it in.

shivaniikum commented:
> Ok, this is looking great! Thanks for making those changes.
> 
> The environment issues were specific to my machine and how I had set things up-- I've also tested on derecho (an NCAR machine) and everything looks good with the environment setup now.
> 
> The new method of finding the configuration file `run_dir` seems to be working smoothly.
> 
> I'll approve this PR, but I think @rmshkv wanted to test it out as well before we merge it in.

That sounds awesome! Thank you!

mnlevy1981 commented on cupid/clear.py:
```diff
@@ -0,0 +1,53 @@
+
+    run_dir = readConfigFile(configFilePath)
+
+    if run_dir:
```
Do we need this `if` statement? If `run_dir` is empty, `readConfigFile()` will raise a `ValueError`

mnlevy1981 commented on cupid/clear.py:
```diff
@@ -0,0 +1,53 @@
+    try:
+        #Obtain the contents of the config.yml file and extract the run_dir variable
+        control = cupid.util.get_control_dict(configFilePath)
+        run_dir = control['data_sources']['run_dir']
```
Since you are going to check to see if `run_dir` exists, I would use the `dict.get()` function and set `run_dir` to `None` or an empty string if it is not in `control['data_sources']`:

```py3
run_dir = control['data_sources'].get('run_dir', None)
```

mnlevy1981 reviewed (changes requested):
Just a few small changes, this looks great!

shivaniikum commented:
Thank you for reviewing my PR  @mnlevy1981. I've modified the clear.py script and README.md as you suggested (and also removed redundant exceptions for the ones I noticed python reports). Please let me know how this looks.

shivaniikum commented:
Thank you for reviewing my PR @mnlevy1981. I've modified the clear.py script and README.md as you suggested (and also removed redundant exceptions for the ones I noticed python reports). Please let me know how this looks.

mnlevy1981 commented:
> Thank you for reviewing my PR @mnlevy1981. I've modified the clear.py script and README.md as you suggested (and also removed redundant exceptions for the ones I noticed python reports). Please let me know how this looks.

Thanks for making those changes! Could you add the `try` / `except` block that you removed from `readConfigFile()` to `get_control_dict()`?

shivaniikum commented:
> > Thank you for reviewing my PR @mnlevy1981. I've modified the clear.py script and README.md as you suggested (and also removed redundant exceptions for the ones I noticed python reports). Please let me know how this looks.
> 
> Thanks for making those changes! Could you add the `try` / `except` block that you removed from `readConfigFile()` to `get_control_dict()`?

Sure, I just made that change.

TeaganKing commented:
I do see the benefit of adding that try/except block to [get_control_dict()](https://github.com/shivaniikum/CUPiD/blob/0cb7d6d6454be0367a36255fecd73c7ea6831bdd/cupid/util.py#L76C5-L76C21) to clarify the file not found error and avoid clutter, since this error may come up somewhat often.

Regarding the `run_dir` error checking, I don't have a strong preference between using a sys.exit and raising a ValueError. It would reduce clutter to use sys.exit, but I think it's still fairly clear that there's a file not found error. My only other concern with skipping the tracebacks would be that if errors occur for unexpected reasons, it could be harder to debug. So, either way is fine by me.

shivaniikum commented:
Thanks for clarifying. I just added the changes you suggested. I agree that these changes make the errors easier to understand.

shivaniikum commented:
I agree that these changes make the errors easier to understand. I've modified util.py and clear.py as suggested by @mnlevy1981.

rmshkv commented:
I think I'd vote for the ValueError rather than sys.exit, looks a little less clean but might be better to keep things consistent.

rmshkv reviewed (approved):
This worked successfully for me! Besides clearing up the last thing Mike mentioned, I think it's good to merge. Thanks so much for your contributions and quick changes @shivaniikum!

mnlevy1981 commented on cupid/clear.py:
```diff
@@ -0,0 +1,47 @@
+import click
+import cupid.util
+
+def clearFolder(folderPath):
+    #Clears all contents in the specified folder at folderPath (i.e. computed_notebooks)
+    try:
+        # Iterate over all items in the folder
+        for item in os.listdir(folderPath):
+            itemPath = os.path.join(folderPath, item)
+            # If item is a file, delete it
+            if os.path.isfile(itemPath):
+                os.remove(itemPath)  
+            # If item is a directory, recursively clear it
+            elif os.path.isdir(itemPath):
+                clearFolder(itemPath)   
+        # After deleting all items, remove the folder itself
+        os.rmdir(folderPath)    
+        print(f"All contents in {folderPath} have been cleared.")
+    except Exception as e:
+        print(f"Error occurred while clearing contents in {folderPath}: {e}")
```
Is this function necessary? You can use [`shutil.rmtree()`](https://docs.python.org/3/library/shutil.html#shutil.rmtree) to remove an entire directory (and its contents).

```diff
-import os
+import shutil
def clear(config_path):
    run_dir = readConfigFile(config_path)
-    clearFolder(run_dir)
+    shutil.rmtree(run_dir)
+    print(f"All contents in {run_dir} have been cleared.")
```

shivaniikum replied:
Thank you for pointing that out. I wasn't aware of the rmtree function. I think rmtree is definitely better than the clearFolder function. I just modified clear.py to use rmtree.

mnlevy1981 commented on cupid/clear.py:
```diff
@@ -0,0 +1,32 @@
+        return fullPath
+    
+    else: #run_dir is empty/wasn't found in config file so return error
+        print("ERROR: 'run_dir' was empty/not found in the config file.")
+        sys.exit(1)
```
I think @TeaganKing and @rmshkv would both prefer this specific `print` / `exit` to be a `raise ValueError` instead (while keeping the `sys.exit()` in `get_control_dict()`), and I'm okay with going back to that.

mnlevy1981 commented on cupid/clear.py:
```diff
@@ -0,0 +1,32 @@
+@click.command()
+@click.argument('config_path')
+#Entry point to this script
+def clear(config_path):
+    run_dir = readConfigFile(config_path)
```
One last minor thing: could you add a docstring to this function so that `cupid-clear --help` describes with the function does?

shivaniikum replied:
I've added both changes. Please let me know how this looks

mnlevy1981 reviewed (approved):
This looks great!

shivaniikum commented:
That's awesome! Thank you so much for reviewing my PR and for all the suggestions @mnlevy1981 @TeaganKing @rmshkv!

TeaganKing commented:
Great work @shivaniikum ! Thank you for contributing this useful feature to the repository!
